Jetpack Partner Portal: Introduce adding credit card as payment method

rcoll opened the pull request:
Design: JVpmiQH74FKEIrPLjqv8p7-fi-1262%3A13812


#### Changes proposed in this Pull Request

* This PR is a port of the composite checkout in the Calypso Green Partner Portal to add credit cards to stripe.
* A lot of this was copied from /me/purchases and adapted to work with Jetpack Cloud. We've trimmed most of what's not needed.
* The PR is functional and the design is almost the same as the proposed, but there are still a few things that can be improved

_Note: this patch currently uses the WPCOM Stripe account, and a later iteration is needed in order to add the separate context for the Payment Methods_

#### Testing instructions

- Check out this PR and run `yarn start-jetpack-cloud`.
- Go to http://jetpack.cloud.localhost:3000/partner-portal/payment-method/add
- Click to add a new card
- Make sure you can go through the process smoothly and that the card is added

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

![image](https://user-images.githubusercontent.com/5714212/125355316-40858580-e33b-11eb-982d-466f71f42862.png)

Commit:
Addressing latest feedback

Commit:
initial commit

Commit:
Partner Portal: Add payment method credit card

a8ci18n commented on client/components/jetpack/add-stored-credit-card/index.tsx:
```diff
@@ -0,0 +1,49 @@
+		>
+			<div className="add-stored-credit-card__content">
+				<CardHeading className="add-stored-credit-card__text" size={ 18 }>
+					{ translate( 'Add new credit card' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'Add a new credit card' )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-partner-portal-add-payment-method-3)

a8ci18n commented on client/components/jetpack/stored-credit-card/index.tsx:
```diff
@@ -0,0 +1,59 @@
+					<EllipsisMenu className="stored-credit-card__ellipsis" position="bottom">
+						<PopoverMenuItem>{ translate( 'Make primary' ) }</PopoverMenuItem>
+						<PopoverMenuItem>{ translate( 'Update details' ) }</PopoverMenuItem>
+						<PopoverMenuItem>{ translate( 'Delete card' ) }</PopoverMenuItem>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'Delete Card', { context: 'Button label'} )` *ES Score: 10*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-partner-portal-add-payment-method-3)

a8ci18n commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -12,10 +20,175 @@ import CardHeading from 'calypso/components/card-heading';
+	);
+
+	if ( paymentMethods.length === 0 ) {
+		return <div>{ translate( 'loading...' ) }</div>;
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **45** times:
`translate( 'Loading…', { context: 'Sharing: Publicize status pending button label'} )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-partner-portal-add-payment-method-3)

IanRamosC committed:
Partner Portal: Fix CSS and trim unnecessary code

a8ci18n commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -11,10 +16,150 @@ import Main from 'calypso/components/main';
+		stripeLoadingError,
+		stripeConfiguration,
+		stripe,
+		activePayButtonText: translate( 'Save payment method' ) as string,
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **13** times:
`translate( 'Saved payment method' )` *ES Score: 8*

kallehauge replied:
@IanRamosC Not really a big thing but we could "speed up the translation and consistency" - as the bot suggests - by just naming it `Save` since the "payment method" part should be understandable in the context of what is happening 😄

IanRamosC committed:
Partner Portal: Remove duplicate component folders

IanRamosC committed:
Partner Portal: Fix add credit card view styling

a8ci18n commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/index.js:
```diff
@@ -0,0 +1,187 @@
+
+	return (
+		<>
+			<CardHeading>{ translate( 'Credit card details' ) }</CardHeading>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'Stored Credit Card Details' )` *ES Score: 9*

IanRamosC committed:
Partner Portal: Add credit card loading placeholder

a8ci18n commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -25,11 +171,86 @@ export default function PaymentMethodAdd(): ReactElement {
+				initiallySelectedPaymentMethodId="card"
+			>
+				<Card className="payment-method-add__content">
+					<CardHeading>{ translate( 'Credit card details' ) }</CardHeading>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **30** times:
`translate( 'Stored Credit Card Details' )` *ES Score: 9*

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/assignment-processor-functions.ts:
```diff
@@ -0,0 +1,193 @@
+	subscriptionId: string,
+	stored_details_id: string
+): Promise< unknown > => wpcom.assignPaymentMethod( subscriptionId, stored_details_id );
+const wpcomCreatePayPalAgreement = (
```
If we wish to keep this as simple as possible should we then remove the `paypal` logic throughout the code?

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/assignment-processor-functions.ts:
```diff
@@ -0,0 +1,193 @@
+	return !! newCardData.name;
+}
+
+interface NewCardSubmitData {
```
We only check the name field in `isNewCardDataValid` but allow for two more arguments here (`countryCode` and `postalCode`)

IanRamosC replied:
This has been removed as the other "Tax Fields" but it's not an issue to add it again

kallehauge commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/style.scss:
```diff
@@ -0,0 +1,36 @@
+			font-weight: 500; /* stylelint-disable-line */
+		}
+	}
+}
```
Missing new line to keep it consistent 😄

kallehauge commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/style.scss:
```diff
@@ -0,0 +1,134 @@
+			@include placeholder( --color-neutral-10 );
+		}
+	}
+}
```
Missing new line to keep it consistent 😄

kallehauge reviewed (commented):
**Some things I noticed:**

_Partner key context_

When you first enter `http://jetpack.cloud.localhost:3000/partner-portal` you're asked to select which key you wish to proceed with (test vs production) but it's the same payment method being used for both. I think it's outside the scope of this PR but it could potentially be something partners could misunderstand.

_Phone number_

When the user adds a payment method we ask for their phone number (if we look past me not knowing why we need this? Money laundering law?) then we never ask for a nationality which means that we work on the assumption that the user will provide a "country prefix" to their phone number. E.g. => `+45`12345678 for a Danish phone.
To give an example of what I mean then it could either be a country select field or like Google and other handle phones and nationality (see attached screenshot)

![Screenshot 2021-07-13 at 13 19 11](https://user-images.githubusercontent.com/3846700/125443263-419646f2-d656-4709-9bc2-c9679cb96794.png)

_Country specific fields_

Normally with Stripe Elements there will be a `Postal` field by default until you start entering in your card details and they magically remove it for non-US users. Is this supported in our solution?
I see references to `postalCode` in different places but never a field for it (the same with `countryCode`, actually).
I'm asking since I cannot test with a US credit card and ensure that it's working. But it works for me with a Danish card.

_Paypal / Credits_

Should we just straight out remove all Paypal and credits specific logic? It doesn't seem like it's something we're considering supporting at the moment because we focus on credit cards and is not guaranteed to work for us when we decide to implement it as payment methods.

_Tests_

The PR doesn't include any tests for the new logic but it's not a dealbreaker since it's not something we strictly do for all of our other Partner Portal components to begin with.
We could, however, consider checking if there's any tests that seems valuable for the credit card specific logic since it's business critical for us to automatically charge for our service.

_Purchases & Setup Intents_

It's not necessary to make a purchase to create a Setup Intent but I guess Blue does it for extra validation of the credit card.
I'm mentioning this since the code could be simplified but I'm not sure it would be a good idea since we plan on adding context support in the wordpress.com/me/ endpoints and they might depend on purchases.

_Strong Customer Authentication_

The SCA modal appeared for me and I could successfully add my credit card after proving to VISA that I am the owner of the card. **Nothing** broke during this process! 👍 

---

The code generally looks like it's been ported (because it has 😅 ). A couple of examples: 

* The code doesn't necessarily look like it comes from the same project because we normally use interfaces more often in the Partner Portal components and some of the terms (like subscriptions) isn't used elsewhere.
* It contains references to variables like `useForAllSubscriptions` even though we don't charge any subscriptions (we will charge Stripe invoices instead).

But until we become wiser from working with the payment endpoints then I don't think we can actually modify much more and should rather add a `Revist Add Payment Method` task which we can update with sub-tasks while working with the next tasks and clean up accordingly since we would just be working on assumptions until then ("that doesn't sound right, let's remove it").

So! It's a functionally working rough version that I feel like is OK for now but will require another go later on when we're wiser 😄

IanRamosC committed:
Partner Portal: Address latest feedback

IanRamosC committed:
Partner Portal: Remove unused code

IanRamosC committed:
Partner Portal: Update stored card component key

kallehauge reviewed (commented):
Awesome work @IanRamosC!

The only thing I'm unsure about before I feel like we should merge this is the postal/tax fields so we know for sure that it works for American customers as well (1187494150150258-as-1200600747566747).

>Country specific fields
>
>Normally with Stripe Elements there will be a Postal field by default until you start entering in your card details and they magically remove it for non-US users. Is this supported in our solution?
>I see references to postalCode in different places but never a field for it (the same with countryCode, actually).
>I'm asking since I cannot test with a US credit card and ensure that it's working. But it works for me with a Danish card.
>
>[Source: André in previous PR comment](https://github.com/Automattic/wp-calypso/pull/53795#pullrequestreview-705066200)

Maybe @oikeme as our American resident can answer if the above sounds like a potential issue if we do not collect the data?

atanas-dev commented:
You're not going to like me for this 😂  but we shouldn't be shipping a 1.5k line patch.
Can we possibly break it down into multiple pieces?

IanRamosC committed:
Partner Portal: Add tests to credit-card store

IanRamosC committed:
Partner Portal: Add the 'credit-card' redux store

This PR adds the base store for the 'credit-card' payment method store that is going to be used in the Partner Portal. It's pretty much a port from what's currently used in the composite-checkout package and might need improvements later.

This is going to be used behind a feature flag, so it won't harm having this merged

IanRamosC committed:
Partner Portal: Remove unused code

IanRamosC committed:
Partner Portal: Fixmissing email and phone in Stripe request

IanRamosC committed:
Partner Portal: Fix condition to set as primary payment method

IanRamosC committed:
Partner Portal: Change the url to handle the credit card

IanRamosC committed:
Partner Portal: Fix method type and call to wrong method name

IanRamosC committed:
Partner Portal: Fix Tracks events when adding payment methods

kallehauge commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/credit-card-submit-button.js:
```diff
@@ -0,0 +1,125 @@
+	const fields = useSelect( ( select ) => select( 'credit-card' ).getFields() );
+	const cardholderName = fields.cardholderName;
+	const cardholderEmail = fields.cardholderEmail;
+	const cardholderPhone = fields.cardholderPhone;
```
Are `cardholderEmail` and `cardholderPhone` valid fields?
Looking at `createStripePaymentMethodStore` in `packages/composite-checkout/dist/esm/lib/payment-methods/stripe-credit-card-fields.js` then I can only find `cardholderName`.

PS: I might be looking in the wrong place though 😅  but it looks like we are dynamically trying to fetch values that I'm not sure is supposed to work like that. E.g.: should we use `fields.getCardholderName()` instead of `fields.cardholderName` ?

kallehauge commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/credit-card-submit-button.js:
```diff
@@ -0,0 +1,125 @@
+	switch ( paymentPartner ) {
+		case 'stripe': {
+			const fields = store.selectors.getFields( store.getState() );
+			const cardholderName = fields.cardholderName;
```
Should we also check  `cardholderEmail` and `cardholderPhone` here?

kallehauge commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/index.js:
```diff
@@ -0,0 +1,194 @@
+				} ) }
+			>
+				<div className="credit-card-fields__form">
+					<Field
```
Just a note: my editor complains about <Field> missing required attributes (description, icon, iconAction, inputClassName, isIconVisible, placeholder, tabIndex) which can be found in `packages/wpcom-checkout/dist/types/field.d.ts`

Edit: this is for the 3 instances of <Field> in this file

kallehauge commented on client/jetpack-cloud/sections/partner-portal/credit-card-fields/style.scss:
```diff
@@ -0,0 +1,134 @@
+	margin-top: 1rem;
+	margin-bottom: 1rem;
+
+	&--is-loaded, &--placeholder {
```
I don't know if it's strictly documented in any "Calypso code standards" but in the rest of the file each new selector is defined on a new line 😄

kallehauge commented on client/jetpack-cloud/sections/partner-portal/index.ts:
```diff
@@ -76,7 +76,7 @@ export default function () {
 
 		page(
-			`/partner-portal/payment-method/add`,
+			`/partner-portal/payment-method/card`,
```
Just for my own edification 😄   Why is it necessary to change this? Will the url for editing a card then be `/partner-portal/payment-method/card/<some_id>/edit` or similar?

kallehauge commented on client/jetpack-cloud/sections/partner-portal/payment-methods/stored-credit-card-method.js:
```diff
@@ -21,13 +21,16 @@ export function createStoredCreditCardMethod( {
 				activeButtonText={ activePayButtonText }
 			/>
 		),
+		// these are not shown in the Partner Portal, but are required for CheckoutProvider
```
I think we typically start comments with a capital letter 🔠 
`// these are not shown in the Partner Portal, but are required for CheckoutProvider`
`// These are not shown in the Partner Portal, but are required for CheckoutProvider`

kallehauge commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -25,11 +111,80 @@ export default function PaymentMethodAdd(): ReactElement {
+				showErrorMessage={ showErrorMessage }
+				showInfoMessage={ showInfoMessage }
+				showSuccessMessage={ showSuccessMessage }
+				paymentMethods={ paymentMethods }
```
Note from my editor: it complains about the `paymentMethods` attribute not accepting `(PaymentMethod | null)` null values but rather `PaymentMethod[]`.

This can be traced back to `const paymentMethods = useMemo( () => [ stripeMethod ].filter( Boolean ), [ stripeMethod ] );` where it thinks `const paymentMethods` isn't a list but rather a single `PaymentMethod`.
I don't think it's the case but worth looking into 😄

kallehauge commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -25,11 +111,80 @@ export default function PaymentMethodAdd(): ReactElement {
 	);
 }
+
+export default function PaymentMethodAddWrapper(): ReactElement {
```
I don't think this is used anywhere? Or is it used later on for something outside this PR? 😄

kallehauge reviewed (changes requested):
**UX Improvement: The tab life**

When I tabbed through the fields while adding my new credit card, I hit the "Go Back" button after the Security Code field which means I accidentally went back and had to re-enter my information again.
I'm not sure if the two buttons should be completely switched around or if we should add `tabindex` but it wasn't the behaviour I expected 😄 

![Screenshot 2021-08-04 at 13 49 50](https://user-images.githubusercontent.com/3846700/128175878-3d26290b-0851-45d5-84ba-01ce677c76a2.png)

**UI Improvement: Loading state**

The loading state for "Add Card" component differs quite a bit from the expected result which means the screen "jumps" a bit after loading.

A couple of things I think we should fix on the loading state are:

1. The space between the "fields"
2. Add label placeholders as well so it vertically align better (this could also just be grey boxes like the fields)

I am not 100% sure how the "Edit card" page will look in terms of fields but if they differ in types of fields then we can add a "How many rows to display" attribute to the loading component so we could say "Show 5 rows/fields" which would then match the "Create fields" experience.
My general understanding of customer experience, then it's less important that we show e.g. two columns like Expiry date and Security Code next to each other compared to the right amount of vertical rows to avoid vertical jumps on load - but even better if we even match the column layout as well - but it should be less important 😄 

![new-card-loading](https://user-images.githubusercontent.com/3846700/128177090-462c2983-b8a5-4dac-8e76-491ba772f6b3.jpg)

**Required fields**

We don't display any indicators for required fields so the partner is first meet with this information when they try to submit their card. Also, are we sure that only the card fields are required here?

![Screenshot 2021-08-04 at 14 10 57](https://user-images.githubusercontent.com/3846700/128178464-4db5537c-7202-4c22-aa09-2a1a3ff22d1c.png)

sirbrillig commented:
👋  I don't know a lot about this project, but please be aware of https://github.com/Automattic/wp-calypso/pull/56089 which changes the Stripe API.

sirbrillig commented on client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:
```diff
@@ -25,11 +111,80 @@ export default function PaymentMethodAdd(): ReactElement {
+						showSuccessMessage,
+					} )
+				}
+				showErrorMessage={ showErrorMessage }
```
Please be aware that the API for `CheckoutProvider` has changed in https://github.com/Automattic/wp-calypso/pull/55809 regarding messages.
